examples/generative/stylegan.py

```diff
@@ -0,0 +1,756 @@
+            1: 512,
+            2: 512,  # 4x4
+            3: 512,  # 8x8
+            4: 512,  # 16x16
+            5: 512,  # 32x32
+            6: 256,  # 64x64
+            7: 128,  # 128x128
+            8: 64,  # 256x256
+            9: 32,  # 512x512
+            10: 16,  # 1024x1024
+        }
+        self.from_rgb = []
+        self.d_blocks = []
+
+        for res_log2 in range(self.start_res_log2, self.target_res_log2 + 1):
+            res = 2 ** res_log2
+            filter_num = self.filter_nums[res_log2]
+            from_rgb = Sequential(
+                [
+                    layers.InputLayer(
+                        input_shape=(res, res, 3), name=f"from_rgb_input_{res}"
+                    ),
+                    EqualizedConv(filter_num, 1),
+                    layers.LeakyReLU(0.2),
+                ],
+                name=f"from_rgb_{res}",
+            )
+
+            self.from_rgb.append(from_rgb)
+
+            input_shape = (res, res, filter_num)
+            if len(self.d_blocks) == 0:
+                d_block = self.build_base(filter_num, res)
+            else:
+                d_block = self.build_block(
+                    filter_num, self.filter_nums[res_log2 - 1], res
+                )
+
+            self.d_blocks.append(d_block)
+
+    def build_base(self, filter_num, res):
+
+        input_tensor = layers.Input(shape=(res, res, filter_num), name=f"d_{res}")
+        x = minibatch_std(input_tensor)
+        x = EqualizedConv(filter_num, 3)(x)
+        x = layers.LeakyReLU(0.2)(x)
+        x = layers.Flatten()(x)
+        x = EqualizedDense(filter_num)(x)
+        x = layers.LeakyReLU(0.2)(x)
+        x = EqualizedDense(1)(x)
+
+        return keras.Model(input_tensor, x, name=f"d_{res}")
+
+    def build_block(self, filter_num_1, filter_num_2, res):
+        input_tensor = layers.Input(shape=(res, res, filter_num_1), name=f"d_{res}")
+
+        x = EqualizedConv(filter_num_1, 3)(input_tensor)
+        x = layers.LeakyReLU(0.2)(x)
+        x = EqualizedConv(filter_num_2)(x)
+        x = layers.LeakyReLU(0.2)(x)
+        x = layers.AveragePooling2D((2, 2))(x)
+
+        return keras.Model(input_tensor, x, name=f"d_{res}")
+
+    def grow(self, res_log2):
+        res = 2 ** res_log2
+        idx = res_log2 - self.start_res_log2
+        alpha = layers.Input(shape=(1), name="d_alpha")
+        input_image = layers.Input(shape=(res, res, 3), name="input_image")
+        x = self.from_rgb[idx](input_image)
+        x = self.d_blocks[idx](x)
+        if idx > 0:
+            idx -= 1
+            downsized_image = layers.AveragePooling2D((2, 2))(input_image)
+            y = self.from_rgb[idx](downsized_image)
+            x = fade_in(alpha[0], x, y)
+
+            for i in range(idx, -1, -1):
+                x = self.d_blocks[i](x)
+
+        return keras.Model([input_image, alpha], x, name=f"discriminator_{res}_x_{res}")
+
+
+"""
+## Build StyleGAN with custom train step
+"""
+
+
+class Phase(Enum):
```

Comment:
You could save a few lines of code by inline string values instead of using an enum (`"transition"` and `"stable"`)

Reply:
that's what I did initially to use string. Then thought it would be a better practise to use enum to avoid spelling mistake.

Reply:
You have a `NotImplementedError`, so there's no worry about that.